truffle/src/main/ruby/core/proc.rb

```diff
@@ -33,18 +41,31 @@ class Proc
 
   def curry(curried_arity = nil)
     if lambda? && curried_arity
+
       if arity >= 0 && curried_arity != arity
-        raise ArgumentError, "Wrong number of arguments (%i for %i)" % [
-          curried_arity,
-          arity
-        ]
+        raise ArgumentError, "Wrong number of arguments (#{curried_arity} for #{arity})"
       end
 
-      if arity < 0 && curried_arity < (-arity - 1)
-        raise ArgumentError, "Wrong number of arguments (%i for %i)" % [
-          curried_arity,
-          -arity - 1
-        ]
+      if arity < 0
+        total_args = 0
+        req_args = 0
+        has_rest = false
+
+        parameters.each do |p|
+          type = p.first
+          req_args += (type == :req ? 1 : 0)
```

Comment:
post `if` is a bit more readable imho: `req_args += 1 if type == :req`,
or alternatively a case:
```ruby
case type
when :req then req_args += 1
...
```